### PR TITLE
[kong] support daemonset deployment for kong pods

### PR DESCRIPTION
**Pull request by arjunkrishnasb**
This commit enables the clients to deploy the kong pods as daemonset, so one instance of pod is deployed in each node.

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ ] PR is based off the current tip of the `next` branch and targets `next`, not `main`
- [ ] New or modified sections of values.yaml are documented in the README.md
- [ ] Title of the PR and commit headers start with chart name (e.g. `[kong]`)

### CODE REVIEW

File: charts/kong/values.yaml
Comment (rainest) [inline charts/kong/values.yaml:599]: If nothing else, this change should be removed. It accidentally reverts https://github.com/Kong/charts/commit/c66fa64dc4944252ff650d042507d08fe24152e9, which we don't want.

**Comment by rainest** [comment]
@arjunkrishnasb can you confirm the rationale for this? Do you have a need to deploy as a DaemonSet for reasons other than handling that particular network configuration?

This seems out of line with what we'd typically expect to support for exposing Kong to clients outside the cluster, and not a standard configuration we'd want to support.

Normally, I'd expect that you either have something that can provide LoadBalancer Services or use a NodePort to have kube-proxy direct inbound traffic to a kubelet hosting a Kong Pod. The latter doesn't provide you with access to reserved ports directly, so if you want that the standard K8S expectation is that you provision a load balancer outside the cluster, have it listen on the standard HTTP/HTTPS ports, and direct traffic to the NodePort. You'd then set [externalIPs](https://kubernetes.io/docs/concepts/services-networking/service/#external-ips) to the load balancer IP to manually provide status information.

I'd recommend [MetalLB](https://metallb.universe.tf/) or similar if your cluster doesn't already have support for LoadBalancer provisioning already.

**Comment by shaneutt** [comment]
@rainest just a heads up, there's relevant context in https://github.com/Kong/kubernetes-ingress-controller/issues/682

**Comment by arjunkrishnasb** [comment]
> @arjunkrishnasb can you confirm the rationale for this? Do you have a need to deploy as a DaemonSet for reasons other than handling that particular network configuration?
> 
> This seems out of line with what we'd typically expect to support for exposing Kong to clients outside the cluster, and not a standard configuration we'd want to support.
> 
> Normally, I'd expect that you either have something that can provide LoadBalancer Services or use a NodePort to have kube-proxy direct inbound traffic to a kubelet hosting a Kong Pod. The latter doesn't provide you with access to reserved ports directly, so if you want that the standard K8S expectation is that you provision a load balancer outside the cluster, have it listen on the standard HTTP/HTTPS ports, and direct traffic to the NodePort. You'd then set [externalIPs](https://kubernetes.io/docs/concepts/services-networking/service/#external-ips) to the load balancer IP to manually provide status information.
> 
> I'd recommend [MetalLB](https://metallb.universe.tf/) or similar if your cluster doesn't already have support for LoadBalancer provisioning already.


That is true. Initially we wanted to ease the deployment as host ports and directly do a TCP mapping to the backend hosts. This makes troubleshooting a bit easier than routing it through nodePort, because that gives introduces kube-proxy at the middle.

Hence i thought we could have daemonset, similar to how nginx offered.

**Comment by debu99** [comment]
when this can be merged?

**Comment by shaneutt** [comment]
> when this can be merged?

It looks like we're still waiting on some review changes and conflict resolution from the original contributor @arjunkrishnasb, once these things are taken care of there are no current blockers so it may be able to merge quickly after that.

Please feel free to share your use case and needs in this matter, end-user context can be helpful to us for priority :vulcan_salute:

File: charts/kong/values.yaml
Comment (rainest) [inline charts/kong/values.yaml:27]: Indentation here is a bit wonky because of the `kong` block above. This applies to the entire replication controller, so it should be only one level below `deployment`

Also making the setting always defined with its default explicitly set. Leaving it undefined by default works since that makes it effectively `false`, but I'd generally prefer to state that explicitly:
```suggestion
  # Deploy as a DaemonSet. By default, the chart creates a Deployment.
  # Toggling this setting to true will use a DaemonSet instead
  daemonset: false
```

**Comment by rainest** [comment]
@arjunkrishnasb I'm still a bit wary of this, but there does appear to be interest here and from issues elsewhere that indicate others want to use DaemonSets, so I suppose we should provide the option.

Do you have a bare metal Kong environment where you've set up a DaemonSet successfully, and did it require additional configuration beyond simply changing the replication controller type? https://kubernetes.github.io/ingress-nginx/deploy/baremetal/#via-the-host-network suggests such an environment would need other settings also.

Added another suggestion to the configuration. Can you add that along reverting the [accidental `smtp_auth_type` change](https://github.com/Kong/charts/pull/269/files#r577838875)?

The longer documentation should be moved to a section under https://github.com/Kong/charts/blob/main/charts/kong/README.md#deployment-options rather than an FAQs.md item. The naming of the FAQs doc is a bit off--in practice we use it to cover things that can go wrong and how to fix them, not for general configuration instructions.

**Comment by arjunkrishnasb** [comment]
Hey Guys. Apologies for the late reply. Let me describe my scenario again. In our environment, k8s does not support loadbalancer type services. So we wanted to use an LB in the front and direct them to the worker nodes. If I expose it using node port, it introduces kubeproxy. But if i expose as node port, then i can just directly do netcat to troubleshoot basic connection and it is just easy for us. So we are deploying it as a daemonset.

This is the same setting we have for another ingress controller in our k8s environment. I am not sure what the use case for others. Based on that kindly repriortize it.

**Review by rainest — CHANGES_REQUESTED** [review]
Can you send in an update with the requested changes to where this covered in documentation (under the README's _Deployment Options_ section rather than the FAQs) and the settings in values.yaml? I do want those at least before merging this.

The other part of my last comment was trying to understand whether deploying a DaemonSet alone handles what you needed or if you were aware of other settings that were necessary also--the NGINX controller docs around this indicate that such environments needed some changes to host network settings. It may be that you can control those with existing chart features, but I wasn't sure, hence why I was curious if you'd done anything else along with changing replication controller type when setting up your environment.

**Comment by rainest** [comment]
Handled in #347

**Pull request closed without merging**
